### PR TITLE
Login/Logout Tests

### PR DESCRIPTION
**Pull request by samuelsandoval1**
PTAL

Had to do this bc got 9000 commits in the other one

### CODE REVIEW

File: src/test/java/com/google/sps/UserLoginTest.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+@RunWith(JUnit4.class)
+public class UserLoginTest {
+
+  @Mock private HttpServletRequest request;
```
Comment (hadleyshapland) [inline src/test/java/com/google/sps/UserLoginTest.java:28]: For style I think @Mock should be on its own line like this:
```
@Mock
private HttpServletRequest request;
```

File: src/test/java/com/google/sps/UserLoginTest.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    when(response.getWriter()).thenReturn(printWriter);
+
+    userServlet = new UserLoginServlet();
+    userServlet.doGet(request, response);
+
+    String responseString = stringWriter.getBuffer().toString().trim();
+    JsonElement responseJsonElement = new JsonParser().parse(responseString);
+
+    JsonObject responseJsonObject = responseJsonElement.getAsJsonObject();
+    String logInUrl = responseJsonObject.get("LogInUrl").getAsString();
```
Comment (hadleyshapland) [inline src/test/java/com/google/sps/UserLoginTest.java:62]: I don't think you use logInUrl in this test, so you don't need to include it

File: src/test/java/com/google/sps/UserLoginTest.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    when(response.getWriter()).thenReturn(printWriter);
+
+    userServlet = new UserLoginServlet();
+    userServlet.doGet(request, response);
+
+    String responseString = stringWriter.getBuffer().toString().trim();
+    JsonElement responseJsonElement = new JsonParser().parse(responseString);
+         
+    JsonObject responseJsonObject = responseJsonElement.getAsJsonObject();
+    String logInUrl = responseJsonObject.get("LogInUrl").getAsString();
```
Comment (hadleyshapland) [inline src/test/java/com/google/sps/UserLoginTest.java:85]: Again, you don't use logInUrl here, so it isn't necessary to include

File: src/test/java/com/google/sps/UserLoginTest.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+    userServlet = new UserLoginServlet();
+    userServlet.doGet(request, response);
+
+    String response = stringWriter.getBuffer().toString().trim();
+    JsonElement responseJsonElement = new JsonParser().parse(response);
+         
+    JsonObject responseJsonObject = responseJsonElement.getAsJsonObject();
+    String logInUrl = responseJsonObject.get("LogInUrl").getAsString();
+    String logOutUrl = responseJsonObject.get("LogOutUrl").getAsString();
```
Comment (hadleyshapland) [inline src/test/java/com/google/sps/UserLoginTest.java:108]: Don't need logOutUrl in this test

File: src/test/java/com/google/sps/UserLoginTest.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+    userServlet = new UserLoginServlet();
+    userServlet.doGet(request, response);
+
+    String response = stringWriter.getBuffer().toString().trim();
+    JsonElement responseJsonElement = new JsonParser().parse(response);
+
+    JsonObject responseJsonObject = responseJsonElement.getAsJsonObject();
+    String logInUrl = responseJsonObject.get("LogInUrl").getAsString();
+    String logOutUrl = responseJsonObject.get("LogOutUrl").getAsString();
```
Comment (hadleyshapland) [inline src/test/java/com/google/sps/UserLoginTest.java:129]: Don't need logOutUrl in this test

File: src/test/java/com/google/sps/UserLoginTest.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    PrintWriter printWriter = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(printWriter);
+
+    userServlet = new UserLoginServlet();
+    userServlet.doGet(request, response);
+
+    String response = stringWriter.getBuffer().toString().trim();
+    JsonElement responseJsonElement = new JsonParser().parse(response);
+
+    JsonObject responseJsonObject = responseJsonElement.getAsJsonObject();
```
Comment (hadleyshapland) [inline src/test/java/com/google/sps/UserLoginTest.java:127]: All four of your tests use the same setup (lines 115-127 in this test). I think you could make this into a helper method that returns the JsonObject to reduce code being repeated in all the tests.

**Comment by jerrychangusa** [comment]
The failure from CloudBuild indicates that this test is currently failing to compile. Could you take a look?

[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /workspace/src/test/java/com/google/sps/UserLoginTest.java:[48,1] illegal start of type
[ERROR] /workspace/src/test/java/com/google/sps/UserLoginTest.java:[48,13] > expected
[ERROR] /workspace/src/test/java/com/google/sps/UserLoginTest.java:[49,3] illegal start of type
[ERROR] /workspace/src/test/java/com/google/sps/UserLoginTest.java:[52,1] illegal start of expression
[ERROR] /workspace/src/test/java/com/google/sps/UserLoginTest.java:[52,3] illegal start of expression

File: src/test/java/com/google/sps/UserLoginTest.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    JsonObject responseJsonObject = jsonSetUp();
+    String logOutUrl = responseJsonObject.get("LogOutUrl").getAsString();
+
+    Assert.assertFalse(logOutUrl.isEmpty());
+    helperLoggedIn.tearDown();
+  }
+
+
+  @Test
+  public void LogOutUrlContainsLogOut() throws ServletException, IOException  {
```
Comment (jerrychangusa) [inline src/test/java/com/google/sps/UserLoginTest.java:76]: Method names should be written in lowerCamelCase.

File: src/test/java/com/google/sps/UserLoginTest.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+    helperLoggedIn.setUp();
+    
+    JsonObject responseJsonObject = jsonSetUp();
+    String logOutUrl = responseJsonObject.get("LogOutUrl").getAsString();
+    
+    Assert.assertTrue(logOutUrl.contains("logout"));
```
Comment (jerrychangusa) [inline src/test/java/com/google/sps/UserLoginTest.java:83]: it seems like if logOutUrl.contains("logout") is true then logOutUrl.isEmpty() would be false by definition. Consider if it's necessary to have two separate tests for verifying the logout url contents.

What might make sense here is to have a separate test that validates that the login url is empty when the user is logged in and that the logout url is empty when the user is logged out.

File: src/test/java/com/google/sps/UserLoginTest.java [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+  private LocalServiceTestHelper helperLoggedIn =
+    new LocalServiceTestHelper(new LocalUserServiceTestConfig())
+        .setEnvIsAdmin(true).setEnvIsLoggedIn(true);
+  
+  private LocalServiceTestHelper helperLoggedOut =
+    new LocalServiceTestHelper(new LocalUserServiceTestConfig())
+        .setEnvIsAdmin(true).setEnvIsLoggedIn(false);
+  
+  private JsonObject jsonSetUp() throws ServletException, IOException {
```
Comment (jerrychangusa) [inline src/test/java/com/google/sps/UserLoginTest.java:42]: maybe name this getLoginServletResponse?

File: src/test/java/com/google/sps/UserLoginTest.java (hunk 7 above)
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/UserLoginTest.java:76]: Done

File: src/test/java/com/google/sps/UserLoginTest.java (hunk 8 above)
Comment (jerrychangusa) [reply src/test/java/com/google/sps/UserLoginTest.java:74]: Is there a reason we need both assertFalse(logOutUrl.isEmpty()) and assertTrue(logOutUrl.contains("logout"))?

File: src/test/java/com/google/sps/UserLoginTest.java [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+@RunWith(JUnit4.class)
+public class UserLoginTest {
+
+  @Mock private HttpServletRequest request;
+
+  @Mock private HttpServletResponse response;
+
+  @Mock private UserLoginServlet userServlet;
+
+  private LocalServiceTestHelper helperLoggedIn =
```
Comment (jerrychangusa) [inline src/test/java/com/google/sps/UserLoginTest.java:34]: would it make sense to use a single LocalServiceTestHelper member variable?

I think you can call setEnvIsLoggedIn on the test helper for each of your tests.

Comment (samuelsandoval1) [reply src/test/java/com/google/sps/UserLoginTest.java]: done

File: src/test/java/com/google/sps/UserLoginTest.java [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+  @Before
+  public void setUp() throws Exception {
+    MockitoAnnotations.initMocks(this);
+  }
+
+  @Test
+  public void userLoggedInReturningLogOut() throws ServletException, IOException  {
+    localHelper.setEnvIsLoggedIn(true);
+    localHelper.setUp();
```
Comment (jerrychangusa) [inline src/test/java/com/google/sps/UserLoginTest.java:62]: since localHelper.setUp() and localHelper.tearDown() need to be called before and after each test, could we move these to setUp and tearDown() methods which are annotated with @Before and @After.

The setup can be called in your existing setUp() method. You will need to create a new tearDown() method which has an @After annotation.

File: src/main/webapp/templates/profileBuilder.html [hunk 12]
```diff
@@ -9,7 +9,8 @@
   <link href="css/styles.css" rel="stylesheet">
 </head>
 <body>
-  
-  <script src="ts/app.ts"></script>
+
+
+  <script src="ts/profileBuilder.ts"></script>
```
Comment (jerrychangusa) [inline src/main/webapp/templates/profileBuilder.html:14]: this looks like an unrelated change from the tests? Can you split this into a separate CL?

Comment (samuelsandoval1) [reply src/main/webapp/templates/profileBuilder.html:14]: yes, sorry about that

File: src/test/java/com/google/sps/UserLoginTest.java (hunk 11 above)
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/UserLoginTest.java]: done

File: src/test/java/com/google/sps/UserLoginTest.java [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+import org.mockito.MockitoAnnotations;
+
+@RunWith(JUnit4.class)
+public class UserLoginTest {
+
+  @Mock private HttpServletRequest request;
+
+  @Mock private HttpServletResponse response;
+
+  @Mock private UserLoginServlet userServlet;
```
Comment (jerrychangusa) [inline src/test/java/com/google/sps/UserLoginTest.java:33]: since you initialize the localHelper, would it make sense to also initialize the userServlet?

File: src/test/java/com/google/sps/UserLoginTest.java [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+  @After
+  public void tearDown() throws Exception {
+    localHelper.tearDown();
+  }
+
+
+  @Test
+  public void userLoggedInReturningLogOut() throws ServletException, IOException  {
+    localHelper.setEnvIsLoggedIn(true);
+    // localHelper.setUp();
```
Comment (jerrychangusa) [inline src/test/java/com/google/sps/UserLoginTest.java:70]: I think this can be removed.

Comment (samuelsandoval1) [reply src/test/java/com/google/sps/UserLoginTest.java:70]: done

File: src/test/java/com/google/sps/UserLoginTest.java [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+  
+  private JsonObject getLoginServletResponse() throws ServletException, IOException {
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter printWriter = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(printWriter);
+
+    userServlet = new UserLoginServlet();
+    userServlet.doGet(request, response);
+
+    String response = stringWriter.getBuffer().toString().trim();
```
Comment (ghandre) [inline src/test/java/com/google/sps/UserLoginTest.java:48]: [L48] You seem to have two variables named 'response' which is slightly confusing. Maybe name this one response_str? Or inline it below in .parse

File: src/test/java/com/google/sps/UserLoginTest.java [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+  private LocalServiceTestHelper localHelper =
+    new LocalServiceTestHelper(new LocalUserServiceTestConfig())
+        .setEnvIsAdmin(true);
+  
+  
+  private JsonObject getLoginServletResponse() throws ServletException, IOException {
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter printWriter = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(printWriter);
+
```
Comment (ghandre) [inline src/test/java/com/google/sps/UserLoginTest.java:44]: Maybe add a one-liner comment before you set up the mock expectations expanding on why you're mocking the printWriter / your testing strategy here. E.g., add the following as L43:
// The UserLoginServlet writes its response to 'response.getWriter()'. Intercept its value and verify.

File: src/test/java/com/google/sps/UserLoginTest.java (hunk 15 above)
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/UserLoginTest.java:48]: done

File: src/test/java/com/google/sps/UserLoginTest.java (hunk 16 above)
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/UserLoginTest.java:44]: Done

File: src/test/java/com/google/sps/UserLoginTest.java (hunk 13 above)
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/UserLoginTest.java]: done

File: src/test/java/com/google/sps/UserLoginTest.java [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+  private JsonObject getLoginServletResponse() throws ServletException, IOException {
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter printWriter = new PrintWriter(stringWriter);
+    /*
+    Mocking the print writer bc
+    UserLoginServlet writes its reponse in response.getWriter() 
+    */
+    when(response.getWriter()).thenReturn(printWriter);
+
+    // UserLoginServlet userServlet = new UserLoginServlet();
```
Comment (jerrychangusa) [inline src/test/java/com/google/sps/UserLoginTest.java:50]: It's best not to have commented out code checked in. maybe remove this?

Comment (samuelsandoval1) [reply src/test/java/com/google/sps/UserLoginTest.java:50]: done
